File: python/src/ahtml/sign.py
"""Detached JWS signatures over canonical snapshot JSON.

Port of ``packages/schema/src/sign.ts`` (see also ``docs/signing.md``).

Wire format: JWS Compact Serialization with a detached payload
(RFC 7515 §3.1, Appendix F) — ``<protected-header>..<signature>``.
The signing input is ``base64url(header) || '.' || base64url(to_json(snap))``
where ``to_json`` is the canonical JSON form of SPEC.md §1.1.

Supported algorithms: ES256, EdDSA (Ed25519), RS256.

WebCrypto (the TS reference) uses raw ``r||s`` ECDSA signatures; the
``cryptography`` package speaks DER, so this module converts between the
two for ES256.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass, field
from typing import Any, Optional, Union

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec, ed25519, padding, rsa
from cryptography.hazmat.primitives.asymmetric.utils import (
    decode_dss_signature,
    encode_dss_signature,
)

from .errors import AHTMLError
from .format_json import to_json
from .snapshot import Snapshot

__all__ = [
    "VerifyKey",
    "SignKey",
    "VerifyResult",
    "import_jwk",
    "sign_snapshot",
    "verify_snapshot",
    "verify_snapshot_strict",
]

SIGN_ALGS = ("ES256", "EdDSA", "RS256")

PublicKey = Union[
    ec.EllipticCurvePublicKey, ed25519.Ed25519PublicKey, rsa.RSAPublicKey
]
PrivateKey = Union[
    ec.EllipticCurvePrivateKey, ed25519.Ed25519PrivateKey, rsa.RSAPrivateKey
]


@dataclass
class VerifyKey:
    """Verifier-side key handle. Verifiers may pass several and try each."""

    alg: str
    key: PublicKey
    kid: Optional[str] = None


@dataclass
class SignKey:
    """Producer-side key handle. ``alg`` MUST match how ``key`` was generated."""

    alg: str
    key: PrivateKey
    kid: Optional[str] = None


@dataclass
class VerifyResult:
    ok: bool
    reason: Optional[str] = None
    signer: dict = field(default_factory=dict)

    def __bool__(self) -> bool:  # pragma: no cover — convenience
        return self.ok


# --------------------------------------------------------------------------
# base64url helpers
# --------------------------------------------------------------------------


def b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def b64url_decode(s: str) -> bytes:
    pad = (4 - len(s) % 4) % 4
    return base64.urlsafe_b64decode(s + "=" * pad)


# --------------------------------------------------------------------------
# JWK import
# --------------------------------------------------------------------------


def _jwk_alg(jwk: dict) -> Optional[str]:
    """Map a JWK to a supported SignAlg. Prefers explicit ``alg``; falls
    back to common kty/crv pairs. Returns None for unsupported keys."""
    alg = jwk.get("alg")
    if alg in SIGN_ALGS:
        return alg
    if jwk.get("kty") == "EC" and jwk.get("crv") == "P-256":
        return "ES256"
    if jwk.get("kty") == "OKP" and jwk.get("crv") == "Ed25519":
        return "EdDSA"
    if jwk.get("kty") == "RSA":
        return "RS256"
    return None


def import_jwk(jwk: dict, *, kid: Optional[str] = None) -> Optional[VerifyKey]:
    """Import a public JWK into a :class:`VerifyKey`.

    Returns None when the key type/curve is unsupported (mirroring the TS
    did:web resolver, which skips such entries rather than throwing).
    """
    alg = _jwk_alg(jwk)
    if alg is None:
        return None
    try:
        if alg == "ES256":
            x = int.from_bytes(b64url_decode(jwk["x"]), "big")
            y = int.from_bytes(b64url_decode(jwk["y"]), "big")
            key: PublicKey = ec.EllipticCurvePublicNumbers(
                x, y, ec.SECP256R1()
            ).public_key()
        elif alg == "EdDSA":
            key = ed25519.Ed25519PublicKey.from_public_bytes(b64url_decode(jwk["x"]))
        else:  # RS256
            n = int.from_bytes(b64url_decode(jwk["n"]), "big")
            e = int.from_bytes(b64url_decode(jwk["e"]), "big")
            key = rsa.RSAPublicNumbers(e, n).public_key()
    except Exception:
        return None
    resolved_kid = kid or jwk.get("kid")
    return VerifyKey(alg=alg, key=key, kid=resolved_kid)


# --------------------------------------------------------------------------
# Sign
# --------------------------------------------------------------------------


def _raw_sign(alg: str, key: PrivateKey, data: bytes) -> bytes:
    if alg == "ES256":
        der = key.sign(data, ec.ECDSA(hashes.SHA256()))
        r, s = decode_dss_signature(der)
        return r.to_bytes(32, "big") + s.to_bytes(32, "big")
    if alg == "EdDSA":
        return key.sign(data)
    if alg == "RS256":
        return key.sign(data, padding.PKCS1v15(), hashes.SHA256())
    raise AHTMLError("SIGNATURE_INVALID", f"unsupported algorithm: {alg}")


def sign_snapshot(
    snap: Snapshot | dict,
    key: SignKey,
    *,
    kid: Optional[str] = None,
    algorithm: Optional[str] = None,
) -> str:
    """Produce a detached JWS over ``to_json(snap)``:
    ``<base64url(header)>..<base64url(signature)>``."""
    alg = algorithm or key.alg
    effective_kid = kid if kid is not None else key.kid

    header: dict[str, Any] = {"alg": alg}
    if effective_kid is not None:
        header["kid"] = effective_kid
    header_b64 = b64url_encode(json.dumps(header, separators=(",", ":")).encode())

    payload_b64 = b64url_encode(to_json(snap).encode("utf-8"))
    signing_input = f"{header_b64}.{payload_b64}".encode("ascii")

    sig = _raw_sign(alg, key.key, signing_input)
    return f"{header_b64}..{b64url_encode(sig)}"


# --------------------------------------------------------------------------
# Verify
# --------------------------------------------------------------------------


def _parse_detached_jws(jws: Any) -> tuple[Optional[dict], Optional[str]]:
    """Returns (parsed, error_reason). ``parsed`` is
    ``{"header_b64", "sig_b64", "header"}``."""
    if not isinstance(jws, str):
        return None, "JWS is not a string"
    parts = jws.split(".")
    if len(parts) != 3:
        return None, "JWS must have three segments"
    header_b64, payload_b64, sig_b64 = parts
    if payload_b64 != "":
        return None, "JWS is not in detached form (payload segment is non-empty)"
    if not header_b64 or not sig_b64:
        return None, "JWS header or signature segment is empty"
    try:
        header = json.loads(b64url_decode(header_b64).decode("utf-8"))
    except Exception as err:
        return None, f"JWS header is not valid base64url JSON: {err}"
    if not isinstance(header, dict):
        return None, "JWS header is not an object"
    return {"header_b64": header_b64, "sig_b64": sig_b64, "header": header}, None


def _raw_verify(alg: str, key: PublicKey, signature: bytes, data: bytes) -> bool:
    try:
        if alg == "ES256":
            if len(signature) != 64:
                return False
            r = int.from_bytes(signature[:32], "big")
            s = int.from_bytes(signature[32:], "big")
            key.verify(encode_dss_signature(r, s), data, ec.ECDSA(hashes.SHA256()))
        elif alg == "EdDSA":
            key.verify(signature, data)
        elif alg == "RS256":
            key.verify(signature, data, padding.PKCS1v15(), hashes.SHA256())
        else:
            return False
        return True
    except InvalidSignature:
        return False


def verify_snapshot(
    snap: Snapshot | dict,
    jws: str,
    *,
    trusted_keys: list[VerifyKey],
) -> VerifyResult:
    """Verify a detached JWS produced by ``sign_snapshot`` /
    ``signSnapshot``. Tries each trusted key in order; first success wins.
    Never raises on a signature mismatch — only on programmer errors
    (no trusted keys supplied)."""
    if not trusted_keys:
        raise ValueError("verify_snapshot requires at least one trusted key")

    parsed, error = _parse_detached_jws(jws)
    if parsed is None:
        return VerifyResult(ok=False, reason=error)

    header = parsed["header"]
    if not isinstance(header.get("alg"), str):
        return VerifyResult(ok=False, reason="JWS header missing alg")

    payload_b64 = b64url_encode(to_json(snap).encode("utf-8"))
    signing_input = f"{parsed['header_b64']}.{payload_b64}".encode("ascii")
    try:
        signature = b64url_decode(parsed["sig_b64"])
    except Exception as err:
        return VerifyResult(ok=False, reason=f"signature is not valid base64url: {err}")

    last_reason = "no trusted key matched"
    for candidate in trusted_keys:
        if candidate.alg != header["alg"]:
            last_reason = f"no trusted key matched alg={header['alg']}"
            continue
        # If both sides declare a kid, require equality. A candidate with
        # no kid is a wildcard for any kid.
        if (
            candidate.kid is not None
            and header.get("kid") is not None
            and candidate.kid != header["kid"]
        ):
            last_reason = f"kid mismatch (header={header['kid']})"
            continue
        try:
            ok = _raw_verify(candidate.alg, candidate.key, signature, signing_input)
        except Exception as err:  # wrong-curve key etc. — try the next key
            last_reason = f"verify threw: {err}"
            continue
        if ok:
            signer: dict[str, Any] = {"alg": candidate.alg}
            if header.get("kid") is not None:
                signer["kid"] = header["kid"]
            elif candidate.kid is not None:
                signer["kid"] = candidate.kid
            return VerifyResult(ok=True, signer=signer)
        last_reason = "signature did not verify"

    return VerifyResult(ok=False, reason=last_reason)


def verify_snapshot_strict(
    snap: Snapshot | dict,
    jws: str,
    *,
    trusted_keys: list[VerifyKey],
) -> Snapshot | dict:
    """Strict variant: raises ``AHTMLError(SIGNATURE_INVALID)`` on failure
    and returns the snapshot unchanged on success."""
    result = verify_snapshot(snap, jws, trusted_keys=trusted_keys)
    if result.ok:
        return snap
    raise AHTMLError(
        "SIGNATURE_INVALID",
        f"snapshot signature verification failed: {result.reason}",
        hint="The snapshot bytes do not match the signature. Re-fetch the "
        "snapshot, and confirm the issuer's did:web document is current.",
    )

File: packages/schema/src/sign.ts
/**
 * v0.8.0 — detached JWS signatures over canonical snapshot JSON.
 *
 * The signing payload is `toJson(snapshot)` — the deterministic,
 * key-ordered JSON form defined in `format-json.ts`. Two semantically
 * identical snapshots serialize byte-identically, so the signature is
 * stable across producers and round-trips.
 *
 * Wire format: JWS Compact Serialization with a detached payload
 * (RFC 7515 §3.1, Appendix F). The middle segment is empty, so a JWS
 * looks like `<protected-header>..<signature>`. Producers ship the
 * snapshot JSON separately (e.g. in the response body) and the JWS in a
 * sidecar header (`X-AHTML-Signature`) or inside
 * `snapshot.provenance.signature`. Verifiers reconstruct the signing
 * input by base64url-encoding the protected header, joining it to the
 * base64url of `toJson(snapshot)` with a `.`, and feeding the resulting
 * bytes to `crypto.subtle.verify`.
 *
 * The whole module talks to `globalThis.crypto.subtle` — with a guarded
 * dynamic `node:crypto` fallback that only runs when the global is missing
 * (bare Node 18) — so it runs unchanged on Cloudflare Workers, Deno, Bun,
 * browsers, and Node ≥ 18.
 */

import type { Snapshot } from './types.js';
import { toJson } from './format-json.js';
import { AHTMLError, DEFAULT_HINTS } from './errors.js';
import { trace } from './otel.js';

/** Supported JWS algorithms. Each maps to a WebCrypto verify parameter. */
export type SignAlg = 'ES256' | 'EdDSA' | 'RS256';

/** Producer-side key handle. The `alg` MUST match how `key` was generated. */
export interface SignKey {
  /** Optional Key ID — round-tripped into the JWS protected header. */
  kid?: string;
  alg: SignAlg;
  key: CryptoKey;
}

/** Verifier-side key handle. Verifiers may pass several and try each. */
export interface VerifyKey {
  /** Optional Key ID — when set, only signatures with a matching `kid` are tried against this key. */
  kid?: string;
  alg: SignAlg;
  key: CryptoKey;
}

export interface SignOptions {
  /** Override `key.kid`. Useful when one key has many identifiers. */
  kid?: string;
  /** Override `key.alg`. Rarely needed. */
  algorithm?: SignAlg;
}

export type VerifyResult =
  | { ok: true; signer: { kid?: string; alg: string } }
  | { ok: false; reason: string };

/* -------------------------------------------------------------------------- */
/* base64url helpers                                                          */
/* -------------------------------------------------------------------------- */

const TEXT_ENCODER = new TextEncoder();

function base64urlEncode(bytes: Uint8Array): string {
  // btoa is available in every WebCrypto-capable runtime (Node ≥ 16, Workers,
  // browsers). Build a binary string then convert to base64url.
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  const b64 = btoa(bin);
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

function base64urlEncodeString(s: string): string {
  return base64urlEncode(TEXT_ENCODER.encode(s));
}

function base64urlDecode(s: string): Uint8Array<ArrayBuffer> {
  // Re-pad and translate alphabet.
  const padLen = (4 - (s.length % 4)) % 4;
  const b64 = s.replace(/-/g, '+').replace(/_/g, '/') + '='.repeat(padLen);
  const bin = atob(b64);
  const buf = new ArrayBuffer(bin.length);
  const out = new Uint8Array(buf);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

/* -------------------------------------------------------------------------- */
/* WebCrypto algorithm descriptors                                            */
/* -------------------------------------------------------------------------- */

function algParams(alg: SignAlg): AlgorithmIdentifier | EcdsaParams | RsaPssParams {
  switch (alg) {
    case 'ES256':
      return { name: 'ECDSA', hash: { name: 'SHA-256' } };
    case 'EdDSA':
      // Node 22 + modern browsers expose Ed25519 directly. Workers may
      // surface it under the same name; the standardized identifier is
      // simply 'Ed25519'.
      return { name: 'Ed25519' };
    case 'RS256':
      return { name: 'RSASSA-PKCS1-v1_5' };
  }
}

let subtleCached: SubtleCrypto | null = null;

async function subtle(): Promise<SubtleCrypto> {
  if (subtleCached) return subtleCached;
  const c = (globalThis as { crypto?: Crypto }).crypto;
  if (c?.subtle) return (subtleCached = c.subtle);
  // Bare Node 18 exposes Web Crypto only behind a flag; load it from
  // node:crypto. Edge runtimes always have the global, so they never reach
  // this import.
  try {
    const { webcrypto } = await import('node:crypto');
    if (webcrypto?.subtle) return (subtleCached = webcrypto.subtle as unknown as SubtleCrypto);
  } catch {
    /* not Node — fall through to the throw below */
  }
  throw new Error('Web Crypto (globalThis.crypto.subtle) is not available in this runtime');
}

/* -------------------------------------------------------------------------- */
/* Sign                                                                       */
/* -------------------------------------------------------------------------- */

/**
 * Produce a detached JWS over `toJson(snap)`.
 *
 * Output format: `<base64url(header)>..<base64url(signature)>` (detached;
 * the payload segment is intentionally empty). Re-running with the same
 * `(snap, key)` produces byte-identical JWS for deterministic algorithms;
 * ECDSA is non-deterministic by design, so successive calls produce
 * different signatures that all verify.
 */
export async function signSnapshot(
  snap: Snapshot,
  key: SignKey,
  opts: SignOptions = {},
): Promise<string> {
  const alg = opts.algorithm ?? key.alg;
  const kid = opts.kid ?? key.kid;

  const headerObj: { alg: SignAlg; kid?: string } = { alg };
  if (kid !== undefined) headerObj.kid = kid;
  const headerJson = JSON.stringify(headerObj);
  const headerB64 = base64urlEncodeString(headerJson);

  const payloadJson = toJson(snap);
  const payloadB64 = base64urlEncodeString(payloadJson);

  const signingInput = TEXT_ENCODER.encode(`${headerB64}.${payloadB64}`);

  const sigBuf = await (await subtle()).sign(algParams(alg), key.key, signingInput);
  const sigB64 = base64urlEncode(new Uint8Array(sigBuf));

  // Detached form: empty payload segment.
  return `${headerB64}..${sigB64}`;
}

/**
 * Detached JWS over arbitrary payload bytes — the same profile
 * `signSnapshot` uses, for non-snapshot payloads (SPEC §4.7 simulated
 * responses, conformance attestations). Additive export (1.1).
 */
export async function signBytes(payload: Uint8Array, key: SignKey, opts: SignOptions = {}): Promise<string> {
  const alg = opts.algorithm ?? key.alg;
  const kid = opts.kid ?? key.kid;
  const headerObj: { alg: SignAlg; kid?: string } = { alg };
  if (kid !== undefined) headerObj.kid = kid;
  const headerB64 = base64urlEncodeString(JSON.stringify(headerObj));
  const payloadB64 = base64urlEncode(payload);
  const signingInput = TEXT_ENCODER.encode(`${headerB64}.${payloadB64}`);
  const sigBuf = await (await subtle()).sign(algParams(alg), key.key, signingInput);
  return `${headerB64}..${base64urlEncode(new Uint8Array(sigBuf))}`;
}

/** Verify a detached JWS produced by {@link signBytes}. Additive (1.1). */
export async function verifyBytes(
  payload: Uint8Array,
  jws: string,
  keys: VerifyKey[],
): Promise<boolean> {
  const parsed = parseDetachedJws(jws);
  if ('error' in parsed) return false;
  const signingInput = TEXT_ENCODER.encode(`${parsed.headerB64}.${base64urlEncode(payload)}`);
  const sig = base64urlDecode(parsed.sigB64);
  for (const key of keys) {
    if (parsed.header.alg && parsed.header.alg !== key.alg) continue;
    try {
      const ok = await (await subtle()).verify(algParams(key.alg), key.key, sig, signingInput);
      if (ok) return true;
    } catch {
      /* try the next key */
    }
  }
  return false;
}

/* -------------------------------------------------------------------------- */
/* Verify                                                                     */
/* -------------------------------------------------------------------------- */

interface ParsedJws {
  headerB64: string;
  sigB64: string;
  header: { alg?: string; kid?: string };
}

function parseDetachedJws(jws: string): ParsedJws | { error: string } {
  if (typeof jws !== 'string') return { error: 'JWS is not a string' };
  const parts = jws.split('.');
  if (parts.length !== 3) return { error: 'JWS must have three segments' };
  const [headerB64, payloadB64, sigB64] = parts as [string, string, string];
  if (payloadB64 !== '') return { error: 'JWS is not in detached form (payload segment is non-empty)' };
  if (!headerB64 || !sigB64) return { error: 'JWS header or signature segment is empty' };
  let header: { alg?: string; kid?: string };
  try {
    const headerText = new TextDecoder().decode(base64urlDecode(headerB64));
    header = JSON.parse(headerText);
  } catch (err) {
    return { error: `JWS header is not valid base64url JSON: ${(err as Error).message}` };
  }
  if (!header || typeof header !== 'object') return { error: 'JWS header is not an object' };
  return { headerB64, sigB64, header };
}

/**
 * Verify a detached JWS produced by `signSnapshot`. Tries each trusted key
 * in order; the first success wins. Returns a `VerifyResult` and NEVER
 * throws on a signature mismatch — only on programmer errors (e.g. no
 * trusted keys supplied, runtime missing Web Crypto).
 */
export async function verifySnapshot(
  snap: Snapshot,
  jws: string,
  opts: { trustedKeys: VerifyKey[] },
): Promise<VerifyResult> {
  // OTel span (no-op when @opentelemetry/api is absent). Also covers
  // verifySnapshotStrict, which delegates here.
  return trace('ahtml.verify_signature', () => verifySnapshotImpl(snap, jws, opts), {
    'ahtml.url': snap?.url,
  });
}

async function verifySnapshotImpl(
  snap: Snapshot,
  jws: string,
  opts: { trustedKeys: VerifyKey[] },
): Promise<VerifyResult> {
  if (!opts || !Array.isArray(opts.trustedKeys) || opts.trustedKeys.length === 0) {
    throw new Error('verifySnapshot requires at least one trusted key');
  }
  const parsed = parseDetachedJws(jws);
  if ('error' in parsed) return { ok: false, reason: parsed.error };

  const { headerB64, sigB64, header } = parsed;
  if (typeof header.alg !== 'string') {
    return { ok: false, reason: 'JWS header missing alg' };
  }

  const payloadB64 = base64urlEncodeString(toJson(snap));
  const signingInput = TEXT_ENCODER.encode(`${headerB64}.${payloadB64}`);
  let signatureBytes: Uint8Array<ArrayBuffer>;
  try {
    signatureBytes = base64urlDecode(sigB64);
  } catch (err) {
    return { ok: false, reason: `signature is not valid base64url: ${(err as Error).message}` };
  }

  let lastReason = 'no trusted key matched';
  for (const candidate of opts.trustedKeys) {
    if (candidate.alg !== header.alg) {
      lastReason = `no trusted key matched alg=${header.alg}`;
      continue;
    }
    // If both sides declare a kid, require equality. If the candidate has
    // no kid, it's a wildcard for any kid.
    if (candidate.kid !== undefined && header.kid !== undefined && candidate.kid !== header.kid) {
      lastReason = `kid mismatch (header=${header.kid})`;
      continue;
    }
    try {
      const ok = await (await subtle()).verify(
        algParams(candidate.alg),
        candidate.key,
        signatureBytes,
        signingInput,
      );
      if (ok) {
        const signer: { kid?: string; alg: string } = { alg: candidate.alg };
        if (header.kid !== undefined) signer.kid = header.kid;
        else if (candidate.kid !== undefined) signer.kid = candidate.kid;
        return { ok: true, signer };
      }
      lastReason = 'signature did not verify';
    } catch (err) {
      // Verify-side errors (wrong-curve key, etc.) are treated like
      // a failed match — try the next key.
      lastReason = `verify threw: ${(err as Error).message}`;
    }
  }

  return { ok: false, reason: lastReason };
}

/**
 * Strict variant: throws `AHTMLError(SIGNATURE_INVALID)` on failure and
 * returns the snapshot unchanged on success. Useful in pipelines that
 * want signature verification to be a hard gate.
 */
export async function verifySnapshotStrict(
  snap: Snapshot,
  jws: string,
  opts: { trustedKeys: VerifyKey[] },
): Promise<Snapshot> {
  const result = await verifySnapshot(snap, jws, opts);
  if (result.ok) return snap;
  throw new AHTMLError({
    code: 'SIGNATURE_INVALID',
    message: `snapshot signature verification failed: ${result.reason}`,
    hint: DEFAULT_HINTS.SIGNATURE_INVALID,
  });
}
